auth/tokencache/tokensource.go

```diff
@@ -0,0 +1,86 @@
+package tokencache
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"time"
+
+	"golang.org/x/oauth2"
+)
+
+// sourceType identities the type of token source.
+type sourceType = string
+
+// cachingTokenSource acts as a read-through cache for token information received from token sources and oauth configurations.
+type cachingTokenSource struct {
+	cacheFile        string
+	login            bool
+	forceLogin       bool
+	sourceType       sourceType
+	sourceIdentifier string
+	oauthTokenSource oauth2.TokenSource
+	oauthConfig      oAuth2Config
+	fetchTimeout     time.Duration
+}
+
+// Token implements the oauth2.TokenSource interface. It will read cached tokens from a file and based on their validity
+// return, refresh or replace them.
+func (source *cachingTokenSource) Token() (*oauth2.Token, error) {
+	// Read the cache information from the file, if it exists
+	cachedTokens, err := readCache(source.cacheFile)
+	if err != nil {
+		log.Println(err)
+	}
+
+	// Garbage collect expired tokens
+	cachedTokens.removeExpiredTokens()
+
+	// Handle login tokens
+	if source.login {
+		return source.loginToken(cachedTokens)
+	}
```

Comment:
I initially had introduced the distinction because I was planning to support service-principal based login. Since that isn't supported right now it is possible to just rely on the `sourceType` and I have refactored the code to do so.